### ISSUE TITLE
Cannot send transaction with insufficient amount even if rectified by earlier transaction in block

### ISSUE THREAD

**Issue by jacob-thrackle** [NomicFoundation/hardhat#4361]
### Version of Hardhat

2.17.2

### What happened?

I am working on replaying historical transactions in a local fork and have come to an impasse. The transaction in question is `0xb368770206f3935625c278cf6c5fabf1776bcbaf59b75f12e4625b4ff10bf3d9` [[link](https://etherscan.io/tx/0xb368770206f3935625c278cf6c5fabf1776bcbaf59b75f12e4625b4ff10bf3d9)] in block 18031132 (the last transaction in the block). When attempting to replay the transaction, we get the following error:
```
ProviderError: sender doesn't have enough funds to send tx. The max upfront cost is: 206908257703903715 and the sender's account only has: 22293390454725773
```

The sender's account (`0x116417a40BcE2bA2acD84d8Da46C60D2F7398D50`) has a balance of 22293390454725773 at the start of the block (end of block 18031131). However, the transaction just before the one in question is `[0x6d0e3c3cfba6dfbd2c62059f80fe89d67d51117463558f0ed2e17a0d457ad22e]`[[link](https://etherscan.io/tx/0x6d0e3c3cfba6dfbd2c62059f80fe89d67d51117463558f0ed2e17a0d457ad22e)] and has an internal transfer of 206908257703903715 to `0x116417a40BcE2bA2acD84d8Da46C60D2F7398D50`. This is the max upfront cost mentioned in the error message.

I am putting the transactions into the "mempool" (automine is false, order is fifo) and executing them together using `evm_mine` in the correct order. The issue is that I am unable to even send transaction `0xb368770206f3935625c278cf6c5fabf1776bcbaf59b75f12e4625b4ff10bf3d9` since the sender has an insufficient amount at the start of the block... but the transaction just before rectifies this. I need to be able to send the transaction anyway but it appears that hardhat blocks this by doing an internal check.

Is there a way to "turn off" this check in hardhat? If not, hardhat *should* have a way to disable these checks to allow transactions such as the above to be added to the queue regardless of the state of the EVM at the end of the last block. My understanding is that since this is possible on mainnet, it should be possible in hardhat as well and I would consider the inability to do so a bug.

### Minimal reproduction steps

1. Fork at block 18031131, turning off automining and using fifo ordering.
2. Build transaction `0x6d0e3c3cfba6dfbd2c62059f80fe89d67d51117463558f0ed2e17a0d457ad22e` using

```javascript
TransactionResponse {
  provider: AlchemyProvider { apiKey: '1ClyQ7h3R0yyONTe9eg915hWsnwXmjAh' },
  blockNumber: 18031132,
  blockHash: '0xb1647833c1576d205202f77e9ca041328542ba11b6077dd8376aaff75010faf2',
  index: undefined,
  hash: '0x6d0e3c3cfba6dfbd2c62059f80fe89d67d51117463558f0ed2e17a0d457ad22e',
  type: 2,
  to: '0x356cfd6e6D0000400000003900b415f80669009e',
  from: '0x0Aa8EBb6aD5A8e499E550ae2C461197624c6e667',
  nonce: 441,
  gasLimit: 41908n,
  gasPrice: 14428330923n,
  maxPriorityFeePerGas: 0n,
  maxFeePerGas: 14428330923n,
  data: '0x000000000b160aa8ebb6ad5a8e499e550ae2c461197624c6e6670aea1c000e6a130000000000000002df15f6d18941e30000116417a40bce2ba2acd84d8da46c60d2f7398d50',
  value: 206908257703903715n,
  chainId: 1n,
  signature: Signature { r: "0x420fb72c24c155a378ead5644ff1f4721abb31c017694478469e8e8ac34089a8", s: "0x1d3326f7d86d9da7d881ad872513f1eae9ccb999978b81c961082a64114bc20d", yParity: 0, networkV: null },
  accessList: []
}
```
3. Build transaction `0xb368770206f3935625c278cf6c5fabf1776bcbaf59b75f12e4625b4ff10bf3d9` using
```javascript
TransactionResponse {
  provider: AlchemyProvider { apiKey: '1ClyQ7h3R0yyONTe9eg915hWsnwXmjAh' },
  blockNumber: 18031132,
  blockHash: '0xb1647833c1576d205202f77e9ca041328542ba11b6077dd8376aaff75010faf2',
  index: undefined,
  hash: '0xb368770206f3935625c278cf6c5fabf1776bcbaf59b75f12e4625b4ff10bf3d9',
  type: 2,
  to: '0x157C9670Cbda36a9046c1f73f86eca2d66077a43',
  from: '0x116417a40BcE2bA2acD84d8Da46C60D2F7398D50',
  nonce: 1271,
  gasLimit: 27300n,
  gasPrice: 14428330923n,
  maxPriorityFeePerGas: 0n,
  maxFeePerGas: 14428330923n,
  data: '0x',
  value: 206514364269705815n,
  chainId: 1n,
  signature: Signature { r: "0x09b1d067daa2029f6392556312f6a7290d6375933acf9112e30377e0a38a9a73", s: "0x31003008a42dd39e24257e57851e361e0081223d3266d77ca0cd4f2fc3a51ede", yParity: 1, networkV: null },
  accessList: []
}
```
4. Call `evm_mine`.
5. Get error.

### Search terms

fork, replay

**Comment by jacob-thrackle** [comment]
I realize now (after reading) that this is default Geth behavior. However, a mining node has the opportunity to reorder transactions. In `fifo` transaction ordering, one should be allowed to play the ordering-miner and not have a restriction on valid/invalid transactions added to the mempool.

**Comment by fvictorio** [comment]
> In `fifo` transaction ordering, one should be allowed to play the ordering-miner and not have a restriction on valid/invalid transactions added to the mempool.

I'm not completely sure. On one hand, we want to mimic these kinds of behavior, so that Hardhat behavior is closer to a real node's behavior. On the other hand, fifo mode is already different from the real behavior, so adding this extra component to it might be acceptable.

I guess it comes down to how useful this behavior would be. Since you mentioned that you were replaying txs from mainnet, that means that those txs were sendable in the first place? I wonder how that happened?

**Comment by jacob-thrackle** [comment]
>  Since you mentioned that you were replaying txs from mainnet, that means that those txs were sendable in the first place? I wonder how that happened?

Yes, my thought is that the sender must have been a node operator and more than that, the node operator who mined that particular block since my understanding is that they can add any transactions to the mempool. At least if they're not using Geth or some variation of Geth that doesn't enforce that condition.

> I guess it comes down to how useful this behavior would be.

It would be useful from my perspective (I've also seen other issues online where people seemed to be trying to do the same). It's also the case that Foundry can do this with [rollfork](https://book.getfoundry.sh/cheatcodes/roll-fork?highlight=rollfork#rollfork).
